### PR TITLE
Collapsible Side Nav

### PR DESCRIPTION
**Pull request by lazytype**
## ✍️ Proposed changes

🎟 _Jira ticket:_ [Collapsible Side Nav](https://jira.mongodb.org/browse/PD-935)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

- [x] I have added my new package to the global tsconfig
- [x] I have added my new package to the Table of Contents on the global README

### CODE REVIEW

File: packages/collapsible-side-nav/README.md
Comment (brookescarlett) [inline packages/collapsible-side-nav/README.md:332]: can omit the path details here

File: packages/collapsible-side-nav/README.md
Comment (brookescarlett) [inline packages/collapsible-side-nav/README.md:407]: ditto

File: packages/collapsible-side-nav/src/SideNav.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNav.tsx:94]: The transition in still feels a little jittery to me, while the transition out does not

File: packages/collapsible-side-nav/src/SideNav.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNav.tsx:132]: Lets add to our keyMap in lib (it seems that keyCode is starting to be deprecated anyway, so when we change how we approach key-detection we'll want to make sure we do this everywhere at once)

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNavGroup.tsx:43]: Should handle truncating too-long group names here as well

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: What is the benefit of making a new set each time? If we've checked that the existing set does not have a path, why can't we just add it to the containedPaths set?

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNavGroup.tsx:104]: ditto here?

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNavGroup.tsx:118]: We recreate this variable everywhere, would it make sense to pass this to the provider as well?

File: packages/collapsible-side-nav/src/SideNavItem.tsx
Comment (brookescarlett) [inline packages/collapsible-side-nav/src/SideNavItem.tsx:169]: should we take an `as` prop here to support more types of navigation?

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: Another question: My understanding was that the purpose of using Sets was to enforce a uniqueness of paths, such that we are able to determine an active Item based on whatever path is passed to the entire Navigation. If that assumption is correct (and perhaps I'm misunderstanding) I have two questions: 

1. Right now, it seems that we're only enforcing this uniqueness within a Group, so duplicate paths can exist outside of the same group
2. Is there a world in which we'd want to support a group that has items that lead to the same page. If that's the case, does it make more sense for Items to accept an active prop rather than have the Navigation component being responsible for determining what's active?

Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: Another argument for perhaps supporting the active prop: In .design (for example) we have three routes that map back to the same Item being active.  I need `/component/checkbox/example` and `/component/checkbox/guideline` both to make the Checkbox Item appear active in the SideNav.

Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: > why can't we just add it to the containedPaths set

this would not trigger a rerender since the updated `containedPaths` would equal the previous one.

> My understanding was that the purpose of using Sets was to enforce a uniqueness of paths

The purpose of using sets here is to be able to quickly look up whether an item exists. We could use plain objects for this too, but the Set API is better-suited for this.

> we have three routes that map back to the same Item being active

I don't see how an `active` prop would add additional value. the existing logic seems to look like
```jsx
<SideNavItem active={item === activePage} ...
```
which would translate to
```jsx
<SideNav currentPath={activePage} ... >
  <SideNavItem path={item} ...

Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: In my case for example I'd like to do: 

```
<SideNav>
   <SideNavItem active={path.includes(componentName)}>{componentName}</SideNavItem>
</SideNav>
```

File: packages/collapsible-side-nav/src/SideNavItem.tsx
Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavItem.tsx:215]: I would want to see a use-case for it. Accepting `as` and using `Box` usually makes things more difficult on the development side.

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: (thank you for the other clarification points)

File: packages/collapsible-side-nav/src/SideNavItem.tsx
Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavItem.tsx:215]: using `Link` for example

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: I think the same can be accomplished by accepting a matching function for `currentPath`, but having something like `path.includes(componentName)` seems more unsound, e.g. `'/components/checkbox'.includes('box')`, compared to normalizing the path, e.g. `normalize('/component/checkbox/example') === '/component/checkbox'`

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:118]: I think the naming is not very clear and would be more clear defined locally since one can look up the definition nearby. I also just prefer not to store redundant data.

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: Accepting a matching function would wfm:
```
<SideNav currentPath={(itemPath) => normalizePath(currentPath) === normalizePath(itemPath}

<SideNavItem path="/component/componentName/example"> 
}
```

Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:93]: If you have `normalizePath` then you wouldn't need a matching function
```
<SideNav currentPath={normalizePath(currentPath)} ...
  <SideNavItem path="/component/checkbox" ...

File: packages/collapsible-side-nav/src/SideNavItem.tsx
Comment (lazytype) [reply packages/collapsible-side-nav/src/SideNavItem.tsx:215]: Fair enough 👍

File: packages/collapsible-side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [reply packages/collapsible-side-nav/src/SideNavGroup.tsx:118]: Makes sense

**Comment by brookescarlett** [comment]
just spoke to @DesignerDave, and it seems that this sidenav should replace our existing, so we don't need a new package here

File: packages/icon/src/glyphs/Apps.svg
Comment (brookescarlett) [inline packages/icon/src/glyphs/Apps.svg:1]: Nit: we should add a changeset for Icon additions

File: packages/side-nav/UPGRADE.md
Comment (brookescarlett) [inline packages/side-nav/UPGRADE.md:1]: nit rm space

File: packages/side-nav/src/CollapseButton.tsx
Comment (brookescarlett) [inline packages/side-nav/src/CollapseButton.tsx:117]: Seems like there's an accessibility issue with these Chevrons. See the accessibility tab on Storybook for more info

File: packages/side-nav/src/SideNav.tsx
Comment (brookescarlett) [inline packages/side-nav/src/SideNav.tsx:95]: Why not setCollapsed here

Comment (brookescarlett) [reply packages/side-nav/src/SideNav.tsx:95]: (Why do we need this outside of an inside of the useEffect hook)

File: packages/side-nav/src/SideNavGroup.tsx
Comment (brookescarlett) [inline packages/side-nav/src/SideNavGroup.tsx:163]: ditto here

File: website/pages/component/side-nav/example.tsx
Comment (brookescarlett) [inline website/pages/component/side-nav/example.tsx:95]: Not seeing the active and disabled props work here

File: packages/side-nav/src/SideNav.tsx
Comment (lazytype) [reply packages/side-nav/src/SideNav.tsx:95]: this is so that the following code we can just write `collapsed` instead of `collapsible && collapsed` everywhere and render the correct thing on this render (and not the next render that would be triggered by the useEffect hook)

File: packages/side-nav/src/SideNav.spec.tsx
Comment (lazytype) [inline packages/side-nav/src/SideNav.spec.tsx:2]: working on writing tests for collapsing groups

File: packages/side-nav/src/SideNav.tsx
Comment (brookescarlett) [reply packages/side-nav/src/SideNav.tsx:95]: I might be missing something but doesn't the ternary above read if it's collapsible then use the initialCollapsed prop and if it's not collapsible set it to false.

Comment (brookescarlett) [reply packages/side-nav/src/SideNav.tsx:95]: we're branching on the same variable in both places

Comment (lazytype) [reply packages/side-nav/src/SideNav.tsx:95]: the argument to `useState` only applies to the initial render

**Review by brookescarlett — APPROVED** [review]
Have been trying to poke holes in this for a bit and I can't find anything else. I'd wait on @DesignerDave's feedback before merging, but I don't have any blocking feedback. 

(The animation looks great, btw.)

File: packages/side-nav/README.md
Comment (DesignerDave) [inline packages/side-nav/README.md:25]: Is this no longer a peer dependency?

**Review by DesignerDave — COMMENTED** [review]
Largely just looked at the end result for this – I'm sure you and Brooke have shaved enough yaks around this 😄 

The animation overall is looking absolutely great. I noticed a few things that might make it feel slightly more polished around the edges.

Few things I noticed:
- There's a noticeable area of the expand / collapse button which doesn't show cursor: pointer.
 - It looks like the underline for sections doesn't transition between states like hover.
 - The animation for opening a group is smooth, but the close is less so. It looks like we may be hiding the items within that group via opacity as the transition begins, and it looks a bit jittery.
 -  When transitioning the side navigation from closed to open, the group expand icons (the chevrons) appear with no transition directly on top of the section icons for a brief moment. Transitioning in the expand icons would make the transition feel vastly smoother.
 - More of a note for @audroue, but the text "Expand sidebar ([)" is showing a keyboard shortcut, but it looks an awful lot like a bug, rather than help text.
 - More of a question for @brookescarlett, but when opening the sidebar with the keyboard shortcut, shouldn't the sidebar become focused?

File: packages/side-nav/README.md
Comment (lazytype) [reply packages/side-nav/README.md:25]: I didn't even notice we had this. Is it worth having? A lot of packages will transitively require it too because they use the Button, Interaction-ring, useUsingKeyboardContext, etc. I definitely failed to add this section to the Select and Copyable readmes

Comment (DesignerDave) [reply packages/side-nav/README.md:25]: It's _probably_ something we should keep around, but missing it in a component or two isn't going to make a huge impact since it's likely everywhere already has this component installed.

@brookescarlett I don't feel strongly about this. You wanna make this call?

Comment (brookescarlett) [reply packages/side-nav/README.md:25]: I think we tried to say that it's a peer dep of every component in the repo. I don't feel strongly one way or another, but do think that we should be consistent.

**Comment by DesignerDave** [comment]
Closing in favor of https://github.com/mongodb/leafygreen-ui/pull/837

**Pull request closed without merging**
